File: js/problems/27-remove-element.js
/**
 * @param {number[]} nums
 * @param {number} val
 * @return {number}
 */
var removeElement = function (nums, val) {
  let cur = 0;

  for (let i = 0; i < nums.length; i++) {
    if (nums[i] !== val) {
      nums[cur] = nums[i];

      cur++;
    }
  }

  return nums;
};

console.log(removeElement([3, 2, 2, 3], 3));

File: .leetcode/27.remove-element.cpp
/*
 * @lc app=leetcode id=27 lang=cpp
 *
 * [27] Remove Element
 */

// @lc code=start
class Solution {
 public:
  int removeElement(vector<int>& nums, int val) {
    /*
      Approach 1: Two Pointers (slow and fast)
      time complexity: O(n)
      space complexity: O(1)
    */

    if (nums.empty()) return 0;

    int slow = 0;

    for (int fast = 0; fast < nums.size() - 1; fast++) {
      if (nums[fast] != val) {
        nums[slow] = nums[fast];
        slow++;
      }
    }

    return slow;
  }
};
// @lc code=end
